torch/ao/quantization/dynamic_tracing/auto_trace.py

```diff
@@ -0,0 +1,554 @@
+                args, kwargs = qstate.op_prepare_before_hook(
+                    func, args, kwargs, first_call, qtensor_id, fqn, parent_module)
+                # forward
+                output = super().__torch_function__(func, types, args, kwargs)
+                # run "after" hook
+                output = qstate.op_prepare_after_hook(
+                    func, output, args, first_call, qtensor_id, parent_module)
+                qstate.mark_cur_op_complete(func)
+            else:
+                output = super().__torch_function__(func, types, args, kwargs)
+
+            # TODO: is this right? Don't really understand this
+            if output is NotImplemented:
+                with torch._C.DisableTorchFunction():
+                    output = func(*args, **kwargs).as_subclass(
+                        QuantizationInterceptionProxy)
+                assert output is not NotImplemented
+
+            return output
+
+        def __repr__(self):
+            return f'QuantizationInterceptionProxy({super().__repr__()})'
+
+        # TODO(future PR): add other math overrides
+
+    class QuantizationInterceptionModule(type(model)):  # type: ignore[misc]
+        """
+        An override of user defined subclass of `nn.Module` to enable
+        dynamic tracing for quantization.
+
+        `cur_module` keeps track of the current module in the stack.
+
+        During the fist call, an `AutoQuantizationState` object is created and
+        attached to each non-leaf modules which we need to check for
+        quantizeable operations.
+
+        We override the `__call__` function to do the following for each
+        module:
+
+        If the module is an op which needs quantization:
+
+        1. calls `_auto_quant_state.validate_cur_op` to validate that
+           the currently seen op is the same as what was recorded during tracing
+        2. calls parent module's `._auto_quant_state.op_prepare_before_hook`
+        3. executes the original module forward
+        4. calls parent module's `_auto_quant_state.op_prepare_after_hook`
+        5. calls `_auto_quant_state.mark_cur_op_complete` to increment
+           the current op index in preparation for the next op
+
+        If the module can contain children ops that need quantization:
+
+        1. calls `_auto_quant_state.inputs_prepare_hook` (not implemented yet)
+        2. executes the original module forward
+        3. calls `_auto_quant_state.outputs_prepare_hook`
+
+        Otherwise, calls the original module forward.
+        """
+
+        __interception_module__ = True
+
+        def __call__(self, *args, **kwargs):
+            new_args = map_aggregate(args, convert_to_interception_proxy)
+            new_kwargs = map_aggregate(kwargs, convert_to_interception_proxy)
+            orig_module_call = torch.nn.Module.__call__
+            orig_nn_sequential_forward = torch.nn.Sequential.forward
+
+            def record_module(self, *args, **kwargs):
```

Comment:
optional nit: can we have a more descriptive name like "_patched_module_call"

Reply:
sure